models/page_views/scratch/bigquery/snowplow_web_page_views_this_run.sql

```diff
@@ -0,0 +1,246 @@
+{{ 
+  config(
+    materialized='table',
+    partition_by = {
+      "field": "start_tstamp",
+      "data_type": "timestamp",
+      "granularity": "day"
+    },
+    cluster_by=["domain_sessionid"],
+    tags=["this_run"]
+  ) 
+}}
+
+with page_view_events as (
+select
+  ev.page_view_id,
+  ev.event_id,
+
+  ev.app_id,
+
+  -- user fields
+  ev.user_id,
+  ev.domain_userid,
+  ev.network_userid,
+
+  -- session fields
+  ev.domain_sessionid,
+  ev.domain_sessionidx,
+
+  -- timestamp fields
+  ev.dvce_created_tstamp,
+  ev.collector_tstamp,
+  ev.derived_tstamp,
+  ev.derived_tstamp as start_tstamp,
+
+  ev.doc_width,
+  ev.doc_height,
+
+  ev.page_title,
+  ev.page_url,
+  ev.page_urlscheme,
+  ev.page_urlhost,
+  ev.page_urlpath,
+  ev.page_urlquery,
+  ev.page_urlfragment,
+
+  ev.mkt_medium,
+  ev.mkt_source,
+  ev.mkt_term,
+  ev.mkt_content,
+  ev.mkt_campaign,
+  ev.mkt_clickid,
+  ev.mkt_network,
+
+  ev.page_referrer,
+  ev.refr_urlscheme,
+  ev.refr_urlhost,
+  ev.refr_urlpath,
+  ev.refr_urlquery,
+  ev.refr_urlfragment,
+  ev.refr_medium,
+  ev.refr_source,
+  ev.refr_term,
+
+  ev.geo_country,
+  ev.geo_region,
+  ev.geo_region_name,
+  ev.geo_city,
+  ev.geo_zipcode,
+  ev.geo_latitude,
+  ev.geo_longitude,
+  ev.geo_timezone ,
+
+  ev.user_ipaddress,
+
+  ev.useragent,
+
+  ev.br_lang,
+  ev.br_viewwidth,
+  ev.br_viewheight,
+  ev.br_colordepth,
+  ev.br_renderengine,
+  ev.os_timezone,
+
+  row_number() over (partition by ev.domain_sessionid order by ev.derived_tstamp) AS page_view_in_session_index,
+
+  -- optional fields, only populated if enabled.
+
+  -- iab enrichment fields: set iab variable to true to enable
+  {{ get_iab_fields(var('snowplow__enable_iab')) }},
```

Comment:
This is super clean. 

It might not be possible, but one to mull over - I wonder if in future these 3 can just be one function, which takes a context name and a list of fields that we care about.

Reply:
It can certainly be done. The decision I made was basically a trade off between dryness and cleanness. If we had 1 generalised function I think we would need to store the array of fields and their datatypes within `page_views_this_run`. They are some pretty chunky arrays so it quickly gets pretty congested